integration-tests/tests/src/tests/observable.ts

```diff
@@ -0,0 +1,848 @@
+          expect(name).equals(eventName, `Realm change event #${c} name didn't match`);
+          expect(schema).deep.equals(expectedChanges.schema, `Realm change event #${c} schema didn't match`);
+        },
+      ),
+    ),
+  );
+  return handle;
+}
+
+function expectObjectNotifications<T>(object: Realm.Object<T>, expectedChangeSets: ObjectChangeSet<T>[]) {
+  const handle = createPromiseHandle();
+  object.addListener(
+    createListenerStub(
+      handle,
+      ...expectedChangeSets.map((expectedChanges, c) => (_: Realm.Object<T>, changes: ObjectChangeSet<T>) => {
+        expect(changes).deep.equals(expectedChanges, `Changeset #${c} didn't match`);
+      }),
+    ),
+  );
+  return handle;
+}
+
+function expectCollectionNotifications(collection: Realm.Collection, expectedChangeSets: CollectionChangeSet[]) {
+  const handle = createPromiseHandle();
+  collection.addListener(
+    createListenerStub(
+      handle,
+      ...expectedChangeSets.map((expectedChanges, c) => (_: Realm.Collection, changes: CollectionChangeSet) => {
+        expect(changes).deep.equals(expectedChanges, `Changeset #${c} didn't match`);
+      }),
+    ),
+  );
+  return handle;
+}
+
+function expectDictionaryNotifications(dictionary: Realm.Dictionary, expectedChangeSets: DictionaryChangeSet[]) {
+  const handle = createPromiseHandle();
+  dictionary.addListener(
+    createListenerStub(
+      handle,
+      ...expectedChangeSets.map((expectedChanges, c) => (_: Realm.Dictionary, changes: DictionaryChangeSet) => {
+        expect(changes).deep.equals(expectedChanges, `Changeset #${c} didn't match`);
+      }),
+    ),
+  );
+  return handle;
+}
+
+describe("Observable", () => {
+  // describe("App", () => {});
+
+  describe("Realm", () => {
+    openRealmBeforeEach({
+      schema: [
+        {
+          name: "Person",
+          properties: {
+            name: "string",
+          },
+        },
+      ],
+    });
+
+    it("is observable", function (this: RealmContext) {
+      expectObservableMethods(this.realm);
+    });
+
+    describe("change", () => {
+      it("calls listener", async function (this: RealmContext) {
+        const completion = expectRealmNotifications(this.realm, "change", [{}]);
+
+        this.realm.write(() => {
+          this.realm.create("Person", { name: "Alice" });
+        });
+
+        await completion;
+      });
+
+      it("removes listeners", async function (this: RealmContext) {
+        const handle = createPromiseHandle();
+
+        const listener = createListenerStub(handle, () => {
+          this.realm.removeListener("change", listener);
+          setImmediate(() => {
+            this.realm.write(() => {
+              this.realm.create("Person", { name: "Bob" });
+            });
+          });
+        });
+
+        this.realm.addListener("change", listener);
+
+        this.realm.write(() => {
+          this.realm.create("Person", { name: "Alice" });
+        });
+
+        await handle;
+      });
```

Comment:
I like the pattern taken with the `expectXNotifications` functions. Could something similar be applied to the removal tests for expecting the removal?

Reply:
I agree that the `await expectXNotifications` reads really good, but I hadn't found a good abstraction for this.
I gave it another go: https://github.com/realm/realm-js/pull/6301